.install/verify_build_deps.sh

```diff
@@ -0,0 +1,302 @@
+# ============================================
+
+# This also serves as the list of supported OSes
+declare -A os_package_checkers=(
+  ["ubuntu"]="check_package_dpkg"
+  ["debian"]="check_package_dpkg"
+  ["rocky"]="check_package_rpm"
+  ["amzn2"]="check_package_yum"
+  ["amzn2023"]="check_package_dnf"
+  ["alpine"]="check_package_apk"
+  ["mariner"]="check_package_tdnf"
+  ["azurelinux"]="check_package_tdnf"
+)
+
+# Early bailout if the OS is not supported
+if [[ -z "${os_package_checkers[$OS]}" ]]; then
+  echo -e "${YELLOW}Error: Unsupported operating system: $OS.${NC}"
+  echo -e "${YELLOW}Abort dependency check.${NC}"
+  exit 1
+fi
+
+# Function to check if a command is available
+check_command() {
+  command -v "$1" &> /dev/null
+}
+
+# ubuntu and debian
+check_package_dpkg() {
+  dpkg -l | grep -q " $1 " || dpkg -l | grep -q " $1:"
+}
+
+# rhel
+check_package_rpm() {
+  rpm -q "$1" &> /dev/null
+}
+
+# amzn2
+check_package_yum() {
+  yum list installed "$1" &> /dev/null
+}
+
+# amzn2023
+check_package_dnf() {
+  dnf list installed "$1" &> /dev/null
+}
+
+# alpine
+check_package_apk() {
+  apk info -e "$1" &> /dev/null
+}
+
+# mariner and azure linux
+check_package_tdnf() {
+  tdnf list installed "$1" &> /dev/null
+}
+
+# ============================================
+# Minimal Version Check Functions
+# ============================================
+
+# Define dependencies that need version checking
+# [<dep>] = "<get_version_function> <check_function> <min_version>"
+declare -A version_checks=(
+  ["gcc"]="get_compiler_version check_gcc_min_version 10"
+  ["g++"]="get_compiler_version check_gpp_min_version 10"
+)
+
+# Function to get the version of a compiler
+get_compiler_version() {
+  local program=$1
+  "$program" -dumpversion 2>/dev/null || echo "unknown"
+}
+
+check_compiler_min_version() {
+  local program=$1
+  local min_version=$2
+  local actual_version=$(get_compiler_version "$program")
+
+  # Extract major version number
+  local actual_major=$(echo "$actual_version" | cut -d. -f1)
+
+  # Compare major versions
+  if [[ $actual_major -ge $min_version ]]; then
+    return 0  # Version requirement met
+  else
+    return 1  # Version requirement not met
+  fi
+}
+
+# Specialized version check functions
+check_gcc_min_version() {
+  local min_version=$1
+  check_compiler_min_version "gcc" "$min_version"
+}
+
+check_gpp_min_version() {
+  local min_version=$1
+  check_compiler_min_version "g++" "$min_version"
+}
+
+# ============================================
+# OS-Specific Dependencies
+# ============================================
+
+# Define common dependencies
+declare -A common_dependencies=(
+  ["make"]="command"       # Verify using command -v
+  ["gcc"]="command"        # Verify using command -v
+  ["g++"]="command"        # Verify using command -v
+  ["python3"]="command"    # Verify using command -v
+  ["cmake"]="command"      # Verify using command -v
+  ["cargo"]="command"      # Verify using command -v
```

Comment:
do any of these need a version check too?

Reply:
yes. gcc/g++ is verified to be at least 10+

Reply:
I installed the default cmake on all platforms (<pack_manager> install cmake) and managed to build.
Considering that failing to meet the version requirement will fail the script and as a result abort make, I don't think we should enforce this unless we ensure that the build process itself will indeed fail.
But finding the minimal version for each OS-arch combination can be a tedious task